rulewerk-vlog/src/main/java/org/semanticweb/rulewerk/reasoner/vlog/VLogFastQueryResultIterator.java

```diff
@@ -0,0 +1,234 @@
+package org.semanticweb.rulewerk.reasoner.vlog;
+
+import java.util.Arrays;
+import java.util.HashMap;
+
+/*
+ * #%L
+ * Rulewerk VLog Reasoner Support
+ * %%
+ * Copyright (C) 2018 - 2020 Rulewerk Developers
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+import org.semanticweb.rulewerk.core.model.api.QueryResult;
+import org.semanticweb.rulewerk.core.model.api.Term;
+import org.semanticweb.rulewerk.core.model.implementation.NamedNullImpl;
+import org.semanticweb.rulewerk.core.reasoner.Correctness;
+import org.semanticweb.rulewerk.core.reasoner.QueryResultIterator;
+import org.semanticweb.rulewerk.core.reasoner.implementation.QueryResultImpl;
+
+import karmaresearch.vlog.NotStartedException;
+import karmaresearch.vlog.VLog;
+
+/**
+ * Iterates trough all answers to a query. An answer to a query is a
+ * {@link QueryResult}. Each query answer is distinct.
+ *
+ * @author Markus Kroetzsch
+ *
+ */
+public class VLogFastQueryResultIterator implements QueryResultIterator {
+
+//	/**
+//	 * Use of Java's LinkedHashMap for implementing a simple LRU cache that is used
+//	 * here for mapping VLog ids to terms.
+//	 * 
+//	 * @author Markus Kroetzsch
+//	 *
+//	 * @param <K>
+//	 * @param <V>
+//	 */
+//	static class SimpleLruMap extends LinkedHashMap<Long, Term> {
+//		private static final long serialVersionUID = 7151535464938775359L;
+//		private int maxCapacity;
+//
+//		public SimpleLruMap(int initialCapacity, int maxCapacity) {
+//			super(initialCapacity, 0.75f, true);
+//			this.maxCapacity = maxCapacity;
+//		}
+//
+//		@Override
+//		protected boolean removeEldestEntry(Map.Entry<Long, Term> eldest) {
+//			return size() >= this.maxCapacity;
+//		}
+//	}
+
+	/**
+	 * Simple cache for finding terms for VLog ids that is optimised for the case
+	 * where ids are inserted in a mostly ordered fashion. An LRU strategy is highly
+	 * ineffective for this as soon as the cache capacity is smaller than the number
+	 * of repeatedly used terms, since the cache entries there are always pushed out
+	 * before being needed again. This implementation will at least cache a maximal
+	 * initial fragment in such cases. It is also faster to write and requires less
+	 * memory.
+	 * 
+	 * @author Markus Kroetzsch
+	 *
+	 */
+	static class OrderedTermCache {
+		final private HashMap<Long, Term> terms = new HashMap<>();
```

Comment:
It might make sense to use `maxCapacity` as the initial capacity of the `HashMap`, to avoid repeatedly growing the map.

Reply:
Not sure if we should have such big maps by default. In the case I worked with, e.g., a cache size of around 78 was enough to have 100% hits upon second encounter of an id. That was for a total of around 1500 ids and around 7 million cache lookups.

Reply:
We could also use something smaller, then. The default for `HashMap` is 16, though, which is probably too small.

Reply:
It will need to grow, indeed, but I doubt that this has a measurable performance impact (the time taken by, say, 10 hashmap space duplications should be very small, and we would probably not initialise at a much larger size either).